src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -1785,8 +1810,20 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
 
                     if (msbuildProject != null)
                     {
-                        msbuildProject.MSBuildNuGetProjectSystem.EndProcessing();
+                        if (isVsBatchStartEventRaised)
+                        {
+                            // raise batch end event only when we raised vs batch start
+                            msbuildProject.MSBuildNuGetProjectSystem.EndProcessing();
+                        }
+
+                        // raise nuget batch end event
+                        if (packageProjectEventArgs != null)
```

Comment:
`msbuildProject.MSBuildNuGetProjectSystem.EndProcessing(); should execute after BatchEnd?.Invoke(this, packageProjectEventArgs);`
what's the reasoning here? i guess both are independent of each other, isn't it?